src/Env.hs

```diff
@@ -0,0 +1,320 @@
+    i         <- identifier
+    arguments <- option [] $ parens (argParser `sepBy1` comma) 
+    --try (parens (argParser `sepBy1` comma)) <|> emptyArgList --option [] $ parens (argParser `sepBy1` comma) --try (parens (argParser `sepBy1` comma)) <|> emptyArgList
+    pos <- getPosition
+    return (TConstr (ConstructorInvoker { nameCons = i, argCons = arguments, constructorInvokerPos = pos }))
+typeVarParser' = do
+    i <- typeVarParser
+    return (TTypeVar i)
+
+argParser, varParser', tParser'  :: Parser Arg
+argParser = try tParser' <|> varParser'
+varParser' = do
+     i <- varParser
+     return (AVar i)
+tParser' = do
+    t <- tParser
+    return (AT t)
+
+kParser, kTypeParser, tParser'' :: Parser K
+kParser = try kTypeParser <|> try tParser''
+kTypeParser = do
+     rword "type"
+     pos <- getPosition
+     return (Ktype (Type { typeName = "type", typePos = pos }))
+tParser'' = do
+    t <- tParser
+    return (KT t)
+
+emptyArgList :: Parser [Arg]
+emptyArgList = do
+  lparen
+  rparen
+  return []
+
+----------------------------------- Typechecker aux functions ------------------------------------------
+
+firsts :: [(a, b)] -> [a]
+firsts xs = [x | (x,_) <- xs]
+
+seconds :: [(a, b)] -> [b]
+seconds xs = [x | (_,x) <- xs]
+
+second :: (a, b) -> b
+second (a, b) = b
+
+checkAndGet :: String -> M.Map String v -> SourcePos -> Either String v
+checkAndGet k m pos = case M.lookup k m of
+  Nothing -> Left ("Error in " ++ (sourcePosPretty pos) ++ " : " ++ k ++ " Doesn't exist in the context \n")
+  Just v ->  Right v
+
+lookUpK :: VarEnv -> Arg -> K
+lookUpK e (AT  (TTypeVar t))  = (Ktype ((typeVarMap e) M.! t)) --(Ktype (Type {typeName = "type", typePos = typeVarPos t }))
+lookUpK e (AT  (TConstr t)) =
+        if (nameCons t) `elem` (declaredNames e)
+        then lookUpK e (AVar (VarConst (nameCons t)))
+        else (Ktype (Type { typeName = "type", typePos = constructorInvokerPos t }))
+lookUpK e (AVar v) = (KT ((varMap e) M.! v))
+
+getTypesOfArgs :: VarEnv -> [Arg] -> [K]
+getTypesOfArgs e args = map (lookUpK e) args
+
+updateEnv :: VarEnv -> (Y, K) -> VarEnv
+updateEnv e (TypeVarY y, Ktype t) = e { typeVarMap = M.insert y t $ typeVarMap e }
+updateEnv e (VarY y, KT t)        = e { varMap = M.insert y t $ varMap e }
+updateEnv e err                   = e { errors = (errors e) ++ "Problem in update: " ++ (show err) ++ "\n" }
+
+addName :: String -> VarEnv -> VarEnv
+addName a e = if (a `elem` (names e))
+              then e {errors = (errors e) ++ ("Name already exsist in the context \n")}
+              else e {names = a : (names e)}
+
+addDeclaredName :: String -> VarEnv -> VarEnv
+addDeclaredName a e = if (a `elem` (declaredNames e))
+                      then e { errors = (errors e) ++ ("Name already exsist in the context \n") }
+                      else e { declaredNames = a : (declaredNames e) }
+
+--------------------------------------- Env Data Types ---------------------------------------
+
```

Comment:
I would move all these definitions up to join the AST definitions, esp. since `VarEnv` is used before it's defined

Reply:
But they are not part of the AST, they should remain where they are